File: 496_next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& findNums, vector<int>& nums) {
        if (findNums.empty()) return vector<int>();
        stack<int> sk;
        unordered_map<int, int> mp;
        for (int i = 0; i != nums.size(); ++i) {
            while (!sk.empty() && sk.top() < nums[i]) {
                mp[sk.top()] = nums[i];
                sk.pop();
            }
            sk.push(nums[i]);
        }
        while (!sk.empty()) {
            mp[sk.top()] = -1;
            sk.pop();
        }
        vector<int> result;
        for (auto num : findNums)
            result.push_back(mp[num]);
        return result;
    }
};

File: 496_next-greater-element-i.py
class Solution(object):
    def nextGreaterElement(self, findNums, nums):
        """
        :type findNums: List[int]
        :type nums: List[int]
        :rtype: List[int]
        """
        d = {}
        st = []
        ans = []
        for x in nums:
            while len(st) and st[-1] < x:
                d[st.pop()] = x
            st.append(x)

        return [d.get(x, -1) for x in findNums]
